docs/src/main/asciidoc/se/injection.adoc

```diff
@@ -0,0 +1,850 @@
+link:{service-registry-base-url}/io/helidon/service/registry/Service.Qualifier.html[`@Service.Qualifier`].
+
+[source,java]
+.Create Blue and Green custom qualifiers
+----
+include::{sourcedir}/se/inject/Qualifier2Example.java[tag=snippet_1, indent=0]
+----
+The `@Blue` and `@Green` annotations serve as our new qualifiers.
+It can now be used to qualify services in the same way as
+link:{service-registry-base-url}/io/helidon/service/registry/Service.Named.html[`@Service.Named`].
+
+[source,java]
+.Custom qualifier Blue and Green usage
+----
+include::{sourcedir}/se/inject/Qualifier2Example.java[tag=snippet_2, indent=0]
+----
+
+Once the services are created and qualified, they can be injected in the same way as before using the following approach
+
+[source,java]
+.Custom qualifier usage on injection point
+----
+include::{sourcedir}/se/inject/Qualifier2Example.java[tag=snippet_3, indent=0]
+----
+
+== Factories
+Let's consider we have a contract named `MyContract`.
+
+The simple case is that we have a class that implements the contract, and that is a service, such as:
+
+[source,java]
+----
+@Service.Singleton
+class MyImpl implements MyContract {
+}
+----
+
+This means that the service instance itself serves as the implementation of the contract. When this service is injected into a dependency injection point, we receive an instance of MyImpl.
+
+However, this approach only works if the contract is an interface and we’re implementing it fully. In some cases, this may not be enough, such as when:
+
+- The instance needs to be provided by an external source.
+- The contract is not an interface.
+- The instance may not always be created (e.g., it is optional).
+
+These challenges can be addressed by implementing one of the factory interfaces supported by the Helidon Service Registry:
+
+- <<Supplier, Supplier>>
+- <<ServicesFactory, ServicesFactory>>
+- <<InjectionPointFactory, InjectionPointFactory>>
+- <<QualifiedFactory, QualifiedFactory>>
+
+=== Supplier
+A factory that supplies a single instance (it can also return `Supplier<Optional<MyContract>>` or `Optional<MyContract>`)
+[source,java]
+.Supplier factory
+----
+include::{sourcedir}/se/inject/FactoryExample.java[tag=snippet_1, indent=0]
+----
+
+=== ServicesFactory
+The link:{service-registry-base-url}/io/helidon/service/registry/Service.ServicesFactory.html[`ServicesFactory`]
+should be used only for its intended purpose — creating zero OR n instances at runtime, each assigned different qualifiers.
+This allows for dynamic and flexible service creation
+while ensuring proper differentiation between instances based on their qualifiers.
+
+[source,java]
+----
+include::{sourcedir}/se/inject/FactoryExample.java[tag=snippet_2, indent=0]
+----
+
+=== QualifiedFactory
+link:{service-registry-base-url}/io/helidon/service/registry/Service.QualifiedFactory.html[`QualifiedFactory`]
+is strictly bound to a specific qualifier type.
+It will get executed only for injection points
```

Comment:
The special case should be used here for `java.lang.Object` as a contract - in such a case, the factory will be used for any contract, as long as the qualifier matches.